### ISSUE TITLE
Add support for BBS

### ISSUE THREAD

**Issue by OR13** [w3c-ccg/vc-http-api#69]
We need a HOLDER http api for "derive proof".

**Comment by OR13** [comment]
@tplooker can you sketch what you would like this to look like in JSON here?

**Comment by tplooker** [comment]
@OR13 yeap sure, note here is what the current reference implementation looks like https://github.com/mattrglobal/jsonld-signatures-bbs/blob/master/src/deriveProof.ts#L28.

And a test invoking the API https://github.com/mattrglobal/jsonld-signatures-bbs/blob/master/__tests__/deriveProof.spec.ts#L30.

**Note** JSONLD-signatures-bbs is a general linked data proof suite it is not VC specific, so we have to consider what additional constraints we have to apply to this general API at the VC layer, e.g how would the API appear in say [vc-js](https://github.com/digitalbazaar/vc-js).

My initial thoughts would be something of the following.

```
/**
 * Derives a proof from a verifiable credential featuring a supported linked data proof
 *
 * @param verifiableCredential A verifiable credential featuring a linked data proof capable of proof derivation
 * @param revealDocument A document of the form of a JSON-LD frame describing the terms to selectively derive from the proof document
 * @param options Options for proof derivation
 */
export const deriveProof = async (
  proofDocument: VerifiableCredential,
  revealDocument: any,
  options: any
): Promise<VerifiableCredential> => {
```

**Comment by OR13** [comment]
Naive solution based on https://w3c-ccg.github.io/vc-http-api/:

```
POST https://example.com/prove/derive { verifiableCredential, revealFrame, options }
```

**Comment by peacekeeper** [comment]
To me, "/prove/credentials" is not very intuitive. Even though the input is a "credential", and something is clearly "proven" when this interface is called, the naming sounds a bit strange (what does it mean to "prove a credential"?).

If you want to follow the "controller" procedural naming style closely, then wouldn't it be something like "/credentials/deriveProof"?

**Comment by OR13** [comment]
@peacekeeper  that would work as well. 

I would love to get consensus on this "style issues" so that decisions like this are easier, I know we struggled to do this last time. perhaps we can sync on this on the next ID&D call?

**Comment by peacekeeper** [comment]
@OR13 I think this is out of scope for the DIF ID WG, see the charter here:

https://github.com/decentralized-identity/org/blob/master/Org%20documents/WG%20documents/DIF_ID_WG_charter_v1.pdf.

Since the VC HTTP APIs are now a CCG work item, maybe we try to schedule a dedicated call for this via the CCG mailing list? It seems there are lots of topics to discuss besides this one, e.g. additional verification checks, the relation to the Universal Wallet API, etc.

**Comment by OR13** [comment]
@peacekeeper makes sense, I tried to start support for your suggestion in the recent PRs.

**Comment by tplooker** [comment]
Following the merge of #73 I think the following endpoint would make sense?

`http://api.example.com/credentials/deriveProof` or `http://api.example.com/credentials/derive`

with a request body of the form 

```
{
   "credential": {.....}, // VC with a proof type supporting derivation
   "frame": {.....} // JSON-LD frame documenting the what elements of the credential to derive
}
```

**Comment by OR13** [comment]
i prefer `http://api.example.com/credentials/derive`

I wonder if there is any future proofing for other formats we should be doing...

**Comment by mavarley** [comment]
Do we need this resolved for plugfest 2021?

**Comment by peacekeeper** [comment]
> Do we need this resolved for plugfest 2021?

I think yes

**Comment by bumblefudge** [comment]
> Do we need this resolved for plugfest 2021?

I'm not sure we "need" this, but I **do** think multiple cohort companies have expressed a willingness to test their implementations against this optional subset of tests as part of their plugathon work if the subset is added in time.  Testees opting **out** of these tests should still count as fully complying with the profile, but companies opting **in** will get an important merit badge.  Since many companies (both in and out of the cohort!) will likely use these tests later in 2021, I would consider it a service to the broader community to get them into the profile.

Should we try to get the relevant PRs in before plugathon feature freeze and reviewed as a courtesy to the cohort companies working towards this goal?  That's a question for the overworked editors to opine on, not me 😉

**Comment by OR13** [comment]
IIRC @tplooker has the ball for this, transmute intends to support his proposal and provide an implementation of it over http.

**Comment by mavarley** [comment]
> IIRC @tplooker has the ball for this, transmute intends to support his proposal and provide an implementation of it over http.

SecureKey is also looking to support this proposal.

**Comment by tplooker** [comment]
Resolved in #128
